### PR TITLE
Issue 48 : Allow password reset in case of forgotten password

### PR DESCRIPTION
**Pull request by rupalkhilari**
Have added functionality for resetting the password, along with the associated tests.
Please let me know in case of any comments.

Thanks!

### CODE REVIEW

**Comment by ashimaathri** [comment]
Thanks @rupalkhilari. This week is going to be pretty busy for me, so I may not be able to look at it immediately. But I'll look at it as soon as possible. Sorry!

**Comment by rupalkhilari** [comment]
No problem at all, @ashimaathri , whenever you get the time. Thanks for informing!

File: approval_frame/forms.py [hunk 1]
```diff
@@ -17,3 +18,22 @@ class RegistrationFormCustom(RegistrationFormUniqueEmail):
         error_messages={'invalid': _("This value may contain only letters, "
             + "numbers and ./+/-/_ characters.")}
     )
+
+class PasswordResetFormCustom(PasswordResetForm):
```
Comment (ashimaathri) [inline approval_frame/forms.py]: I'm not sure why we need this class. Isn't everything covered in the PasswordResetForm class?

**Comment by ashimaathri** [comment]
Thanks a lot @rupalkhilari! The changes look really good.

File: approval_frame/forms.py (hunk 1 above)
Comment (rupalkhilari) [inline approval_frame/forms.py]: Actually, in the default PasswordResetForm, I could not find functionality where the form would error out in case the specific email address was not registered with the system. Instead, it would simply not send the email to that address.
So I added it to serve this purpose.  Le me know what you think.
Thanks!

Comment (ashimaathri) [inline approval_frame/forms.py]: I agree that providing an error message to the user would provide a good user experience. Usually Django follows best practices and I was wondering why they haven't included this feature. So, I dug around a bit and found [this answer](http://stackoverflow.com/questions/22582066/why-does-password-reset-works-with-unregistered-email-in-django) on stack overflow. It looks like the absence of this feature is intentional for security reasons.

More info about best practices for a secure password reset [here] (https://www.fishnetsecurity.com/6labs/resource-library/white-paper/best-practices-secure-forgot-password-feature)

TL;DR Password reset application form should not indicate if the username was valid or not and we shouldn't send the username in the password reset email.

Comment (rupalkhilari) [inline approval_frame/forms.py]: Ah, I understand. Nice article. I guess the user in that case should remember the email ID with which he created the account. 
I had noticed that Github does indicate if the email address is a registered one as a part of its validation, so thought that our application should have the same. 
I'll make the required changes though ! Thanks

Comment (ashimaathri) [inline approval_frame/forms.py]: I didn't know that about Github! Looks like they're not following good security measures.
And you've already added the feature that allows them to login using their email ID :smile: 
So we don't have to worry about them forgetting their username.

Comment (rupalkhilari) [inline approval_frame/forms.py]: Sure @ashimaathri , that sounds good. I have removed this validation in the latest commit. Let me know if it is satisfactory. Thanks!

File: approval_frame/forms.py [hunk 2]
```diff
@@ -17,3 +17,4 @@ class RegistrationFormCustom(RegistrationFormUniqueEmail):
         error_messages={'invalid': _("This value may contain only letters, "
             + "numbers and ./+/-/_ characters.")}
     )
+
```
Comment (ashimaathri) [inline approval_frame/forms.py]: I think this got left in by mistake

File: approval_frame/templates/registration/login.html [hunk 3]
```diff
@@ -28,6 +28,11 @@
           <input type='password' class='form-control' id='{{ form.password.id_for_label }}'
           name='{{ form.password.name }}' placeholder='{{ form.password.label }}'>
         </div>
+        <div class='form-group'>
```
Comment (ashimaathri) [inline approval_frame/templates/registration/login.html]: Could you add a `small` class here, it makes the text smaller?

File: approval_frame/urls.py [hunk 4]
```diff
@@ -11,5 +10,7 @@
     url(r'^approval_polls/', include('approval_polls.urls', namespace="approval_polls")),
     url(r'^admin/', include(admin.site.urls)),
     url(r'^accounts/register/$', RegistrationView.as_view(form_class=RegistrationFormCustom), name='registration_register'),
+    url(r'^accounts/', include('registration.auth_urls')),
     url(r'^accounts/', include('registration.backends.default.urls'))
```
Comment (ashimaathri) [inline approval_frame/urls.py]: Do we need both routes (line 13 and 14)?

File: approval_frame/templates/registration/password_reset_complete.html [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{% extends "base.html" %}
+
+{% block content %}
+  <div class='row-fluid top-buffer'>
+    <div class='col-md-12'>
+    <legend>Password reset complete</legend>
+      <p>Your password has been reset. You can now <a href="{% url 'auth_login' %}">login</a>.</p>
+    </div>
+  </div>
+{% endblock %}
```
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_complete.html]: Could you add a `{% block login %}{% endblock %}` at the end to remove the extra login link?

File: approval_frame/templates/registration/password_reset_confirm.html [hunk 6]
```diff
@@ -0,0 +1,61 @@
+{% extends "base.html" %}
+
+{% block content %}
+
+  {% if validlink %}
+    <div class='row-fluid top-buffer'>
+      <div class='col-md-6 col-md-offset-3'>
+        <legend>Set a new password</legend>
```
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_confirm.html]: Could you put the legend tag under the form tag?

File: approval_frame/templates/registration/password_reset_confirm.html [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+          <a href="{% url 'auth_login' %}">Login</a>
+        </p>
+      </div>
+    </div>
+  {% else %}
+    <div class='row-fluid top-buffer'>
+      <div class='col-md-12'>
+        <legend>Password reset unsuccessful</legend>
+        <p>
+          The password reset link was invalid, possibly because it has already been used.
```
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_confirm.html]: I think putting it in the present tense would be more appropriate here 'This password reset link is invalid'

File: approval_frame/templates/registration/password_reset_form.html [hunk 8]
```diff
@@ -0,0 +1,41 @@
+{% extends "base.html" %}
+
+{% block content %}
+    <div class='row-fluid top-buffer'>
+      <div class='col-md-6 col-md-offset-3'>
+        <legend>Reset Password</legend>
+        <p>Enter your email address below and we will send you an email with
```
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_form.html]: Both legend and <p> can go inside the form. The instruction can be shortened to 
'Instructions to reset your password will be sent to the email address entered below.'

File: approval_frame/templates/registration/password_reset_form.html [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            {% for error in form.errors.email %}
+              <label class='control-label' for='{{ form.email.id_for_label }}'>
+                {{ error }}
+              </label>
+            {% endfor %}
+            <input type='text' class='form-control' id='{{ form.email.id_for_label }}'
+                  name='{{ form.email.name }}' placeholder='Email ID'
+                  maxlength={{ form.email.field.max_length }}>
+          </div>
+          <button type='submit' class='btn btn-success'>Submit</button>
```
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_form.html]: Could you rename the button to 'Send reset link'

File: approval_frame/templates/registration/password_reset_form.html [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+         </form>
```
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_form.html]: (Nitpick) Everything from the closing form tag to endblock needs to be dedented to the left by one indent

File: approval_frame/urls.py (hunk 4 above)
Comment (rupalkhilari) [inline approval_frame/urls.py]: I believe we don't actually need both. I missed out the fact that registration.auth_urls was included at the end in registration.backends.default.urls.

File: approval_frame/templates/registration/password_reset_complete.html (hunk 5 above)
Comment (rupalkhilari) [inline approval_frame/templates/registration/password_reset_complete.html]: Sure, I realized that would make it consistent with the activation_complete.html.

File: approval_frame/templates/registration/password_reset_form.html (hunk 9 above)
Comment (rupalkhilari) [inline approval_frame/templates/registration/password_reset_form.html]: I was attempting to conceal the fact that a link would be sent to the email address (for security purposes) by specifying "instructions" in the prompt. If that is not required, I could change both the prompt to "A link to reset your password will be sent to the email address entered below." and subsequently the button "Send reset link"

File: approval_frame/templates/registration/password_reset_form.html (hunk 10 above)
Comment (rupalkhilari) [inline approval_frame/templates/registration/password_reset_form.html]: Hmm, I believe it would be only the {% csrf_token %} that would require indentation ?

File: approval_frame/templates/registration/password_reset_form.html (hunk 8 above)
Comment (rupalkhilari) [inline approval_frame/templates/registration/password_reset_form.html]: Sure, will change that accordingly!

**Comment by rupalkhilari** [comment]
Thanks for your comments @ashimaathri ! Pretty minor, my only doubt was whether we should have "instructions" or an explicit "reset link" mentioned to the user upon reset.

File: approval_frame/templates/registration/password_reset_confirm.html (hunk 7 above)
Comment (rupalkhilari) [inline approval_frame/templates/registration/password_reset_confirm.html]: Thats true, sounds better!

File: approval_frame/templates/registration/password_reset_form.html (hunk 9 above)
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_form.html]: Of course, that makes more sense. I should have said 'email' instead of 'link'.

File: approval_frame/templates/registration/password_reset_form.html (hunk 10 above)
Comment (ashimaathri) [inline approval_frame/templates/registration/password_reset_form.html]: I get it now, the closing form tag was only one space to the left of the button tag above it, so that confused me. Everything is in the right place except for the closing form tag.

File: approval_frame/urls.py (hunk 4 above)
Comment (ashimaathri) [inline approval_frame/urls.py]: Ok, cool. I didn't know if backend.default.urls had all the routes covered by registration.auth_urls, so I just wanted to confirm. Thanks!
